mozilla-mobile/android-components
# Extract the Glean 'metrics.yaml' file from the GV AAR in engine-gecko-nightly

This stack adds the following mechanics to support GeckoView metrics to the Glean SDK and engine-gecko-nightly:

- introduces an option to make `sdk_generator.gradle` look for `metrics.yaml` in AAR files (**note**: this doesn't restrict fetching the files to the geckoview AAR - could this be a problem?);
- ensures the `metrics.yaml` from AAR is available for generating the Kotlin APIs;
- changes the script so that the generation task runs even if no local file is available, as long as the AAR option is enabled;
- it integrates Glean to engine-gecko-nightly as a compileOnly dependency.

**How does this work?**

When the `allowMetricsFromAAR` option is enabled, the Glean SDK build script registers an artifact transformation class (i.e. `GleanMetricsYamlTransform`) that, as per the `registerTransform` block, is applied on an "exploded AAR" (unzipped). It looks for the 'metrics.yaml' in the root dir of the exploded AAR. That file is fed as input to the `generateMetricsSourceFor*` within the `doFirst` block of that task.

**How did I test this?**

- Look up the cached GeckoView AAR file available locally (e.g. `C:\Users\Alessio\.gradle\caches\modules-2\files-2.1\org.mozilla.geckoview\geckoview-nightly\70.0.20190722093634\b4070cbe52a9ae4b8f245dae0bf419564c3cc3d7\geckoview-nightly-70.0.20190722093634.aar` on my machine);
- add a sample `metrics.yaml` file to its root (for example, the one attached [here](https://github.com/mozilla-mobile/android-components/files/3423095/metrics-original.zip))
- Build engine-gecko-nightly
- A `GeckoviewGfx.kt` file should be available in `android-components\components\browser\engine-gecko-nightly\build\generated\source\glean\debug\kotlin`


### Pull Request checklist

- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
